File: Linked_List/Linked_List.py
class node:
    def __init__(self, data):
        self.data = data
        self.next = None

class Linked_List:
    def __init__(self):
        self.head = None

    def Is_List_Empty(self):
        if self.head == None:
            return True

        return False

    def Insert_At_Beginning(self, value):
        temp = node(value)
        temp.next = self.head
        self.head = temp

    def Insert_At_End(self, value):
        temp = node(value)
        temp.next = None

        if self.Is_List_Empty():
            self.head = temp
            return

        current = self.head

        while current.next != None:
            current = current.next

        current.next = temp

    def Insert_After_Value(self, desired, value):
        current = self.head

        while current != None and current.data != desired:
            current = current.next

        if current == None:
            print("Element " + str(desired) + " not in list")
        else:
            temp = node(value)
            temp.next = current.next
            current.next = temp

    def Delete_At_Beginning(self):
        if self.Is_List_Empty():
            print("List is empty")
        else:
            temp = self.head
            self.head = self.head.next
            temp.next = None

    def Delete_At_End(self):
        if self.Is_List_Empty():
            print("List is empty")
            return

        temp = self.head

        if self.head.next == None:
            self.head = None
            temp.next = None
            return

        while temp.next != None:
            prev = temp
            temp = temp.next

        prev.next = temp.next
        temp.next = None

    def Delete_With_Value(self, desired):
        if self.Is_List_Empty():
            print("List is empty")
            return

        temp = self.head

        if self.head.data == desired:
            self.head = self.head.next
            temp.next = None
            return

        while temp != None and temp.data != desired:
            prev = temp
            temp = temp.next

        if temp == None:
            print("Element " + str(desired) + " is not in list")
        else:
            prev.next = temp.next
            temp.next = None


    def Search(self, desired):
        temp = self.head

        while temp != None and temp.data != desired:
            temp = temp.next

        if temp == None:
            print("Element " + str(desired) + " not found")
        else:
            print("Element " + str(desired) + " is present in list")

    def Print_Linked_List(self):
        if self.Is_List_Empty():
            print("List is Empty")
            return

        current = self.head

        while current.next != None:
            print(current.data, end = " -> ")
            current = current.next

        print(current.data)

    def Get_Length_Iterative(self):
        current = self.head
        length = 0

        while current != None:
            length += 1
            current = current.next

        return length

    def Get_Length_Recursive(self):
        def Count_Recursive(node):
            if node is None:
                return 0
            else:
                return 1 + Count_Recursive(node.next)

        return Count_Recursive(self.head)



LinkedList = Linked_List()
print("List Created!")
LinkedList.Print_Linked_List()
print("Current Length:", LinkedList.Get_Length_Iterative())

for i in range(0, 5):
    LinkedList.Insert_At_Beginning(i)

LinkedList.Print_Linked_List()
print("Current Length:", LinkedList.Get_Length_Iterative())

for i in range(5, 10):
    LinkedList.Insert_At_End(i)

LinkedList.Print_Linked_List()
print("Current Length:", LinkedList.Get_Length_Recursive())

LinkedList.Insert_After_Value(5, 9)
LinkedList.Insert_After_Value(10, 9)

LinkedList.Print_Linked_List()
print("Current Length:", LinkedList.Get_Length_Iterative())

for i in range(0, 3):
    LinkedList.Delete_At_End()

LinkedList.Print_Linked_List()
print("Current Length:", LinkedList.Get_Length_Recursive())

for i in range(0, 3):
    LinkedList.Delete_At_Beginning()

LinkedList.Print_Linked_List()
print("Current Length:", LinkedList.Get_Length_Iterative())

LinkedList.Delete_With_Value(1)
LinkedList.Delete_With_Value(5)

LinkedList.Print_Linked_List()
print("Current Length:", LinkedList.Get_Length_Recursive())

LinkedList.Search(6)
LinkedList.Search(8)


'''
Output

List Created!
List is Empty
Current Length: 0
4 -> 3 -> 2 -> 1 -> 0
Current Length: 5
4 -> 3 -> 2 -> 1 -> 0 -> 5 -> 6 -> 7 -> 8 -> 9
Current Length: 10
Element 10 not in list
4 -> 3 -> 2 -> 1 -> 0 -> 5 -> 9 -> 6 -> 7 -> 8 -> 9
Current Length: 11
4 -> 3 -> 2 -> 1 -> 0 -> 5 -> 9 -> 6
Current Length: 8
1 -> 0 -> 5 -> 9 -> 6
Current Length: 5
0 -> 9 -> 6
Current Length: 3
Element 6 is present in list
Element 8 not found

'''

File: Linked_List/Linked_List.java
class Linked_List
{
    node head;

    static class node
    {
        int data;
        node next;

        // Constructor
        node(int d)
        {
            data = d;
            next = null;
        }
    }

    public boolean Is_List_Empty()
    {
        if(head == null)
            return true;

        return false;
    }

    public void Insert_At_Beginning(int value)
    {
        node temp = new node(value);
        temp.next = head;
        head = temp;
    }

    public void Insert_At_End(int value)
    {
        node temp = new node(value), current = head;
        // temp.next = null;

        if(Is_List_Empty())
        {
            head = temp;
            return ;
        }

        while(current.next != null)
            current = current.next;

        current.next = temp;
    }

    public void Insert_After_Value(int desired, int value)
    {
        node current = head;

        while(current != null && current.data != desired)
            current = current.next;

        if(current == null)
            System.out.println("Element " + desired + " is not in list");
        else
        {
            node temp = new node(value);
            temp.next = current.next;
            current.next = temp;
        }
    }

    public void Delete_At_Beginning()
    {
        if(Is_List_Empty())
            System.out.println("List is empty");
        else
        {
            node temp = head;
            head = head.next;
            temp.next = null;
        }
    }

    public void Delete_At_End()
    {
        if(Is_List_Empty())
        {
            System.out.println("List is empty");
            return ;
        }

        node temp = head, prev = null;

        if(head.next == null)
        {
            head = null;
            temp.next = null;
            return ;
        }

        while(temp.next != null)
        {
            prev = temp;
            temp = temp.next;
        }

        prev.next = temp.next;
        temp.next = null;
    }

    public void Delete_With_Value(int desired)
    {
        if(Is_List_Empty())
        {
            System.out.println("List is empty");
            return ;
        }

        node temp = head, prev = null;

        if(head.data == desired)
        {
            head = head.next;
            temp.next = null;
            return ;
        }

        while(temp != null && temp.data != desired)
        {
            prev = temp;
            temp = temp.next;
        }

        if(temp == null)
            System.out.println("Element " + desired + " is not in list");
        else
        {
            prev.next = temp.next;
            temp.next = null;
        }
    }

    public void Search(int desired)
    {
        node temp = head;

        while(temp != null && temp.data != desired)
            temp = temp.next;

        if(temp == null)
            System.out.println("Element " + desired + " not found");
        else
            System.out.println("Element " + desired + " is present in list");
    }

    public void Print_Linked_List()
    {
        if(Is_List_Empty())
        {
            System.out.println("List is empty");
            return ;
        }

        node current = head;

        while(current.next != null)
        {
            System.out.print(current.data + " -> ");
            current = current.next;
        }

        System.out.println(current.data);
    }

    public void Get_Length_Iterative()
    {
        int elementCounter = 0;
        node current = head;

        while (current != null)
        {
            elementCounter++;
            current = current.next;
        }

        System.out.println("Current Length: " + elementCounter);
    }

    public void Get_Length_Recursive()
    {
        System.out.println("Current Length: " + Count_Recursive(head));
    }

    public int Count_Recursive(node currNode)
    {
        if (currNode == null)
            return 0;
        else
            return Count_Recursive(currNode.next) + 1;
    }

    public static void main(String[] args)
    {
        Linked_List LinkedList = new Linked_List();

        int i;

        for(i = 0; i < 5; i++)
            LinkedList.Insert_At_Beginning(i);



        LinkedList.Print_Linked_List();
        LinkedList.Get_Length_Iterative();

        for(i = 5; i < 10; i++)
            LinkedList.Insert_At_End(i);

        LinkedList.Print_Linked_List();
        LinkedList.Get_Length_Recursive();

        LinkedList.Insert_After_Value(5, 9);
        LinkedList.Insert_After_Value(10, 9);

        LinkedList.Print_Linked_List();
        LinkedList.Get_Length_Iterative();

        for(i = 0; i < 3; i++)
            LinkedList.Delete_At_End();

        LinkedList.Print_Linked_List();
        LinkedList.Get_Length_Recursive();

        for(i = 0; i < 3; i++)
            LinkedList.Delete_At_Beginning();

        LinkedList.Print_Linked_List();
        LinkedList.Get_Length_Iterative();

        LinkedList.Delete_With_Value(1);
        LinkedList.Delete_With_Value(5);

        LinkedList.Print_Linked_List();
        LinkedList.Get_Length_Recursive();

        LinkedList.Search(6);
        LinkedList.Search(8);
    }
}

/*  Output
    4 -> 3 -> 2 -> 1 -> 0
    Current Length: 5
    4 -> 3 -> 2 -> 1 -> 0 -> 5 -> 6 -> 7 -> 8 -> 9
    Current Length: 10
    Element 10 is not in list
    4 -> 3 -> 2 -> 1 -> 0 -> 5 -> 9 -> 6 -> 7 -> 8 -> 9
    Current Length: 11
    4 -> 3 -> 2 -> 1 -> 0 -> 5 -> 9 -> 6
    Current Length: 8
    1 -> 0 -> 5 -> 9 -> 6
    Current Length: 5
    0 -> 9 -> 6
    Current Length: 3
    Element 6 is present in list
    Element 8 not found
*/
